packages/ra-ui-materialui/src/field/ArrayField.tsx

```diff
@@ -88,78 +117,40 @@ const initialState = {
  *     )
  *     TagsField.defaultProps = { addLabel: true };
  */
-export class ArrayField extends Component<
-    FieldProps & InjectedFieldProps,
-    State
-> {
-    constructor(props: FieldProps & InjectedFieldProps) {
-        super(props);
-        this.state = props.record
-            ? this.getDataAndIds(props.record, props.source, props.fieldKey)
-            : initialState;
-    }
+export const ArrayField: FunctionComponent<
+    FieldProps & InjectedFieldProps & State
+> = ({
+    addLabel,
+    basePath,
+    children,
+    record,
+    sortable,
+    source,
+    fieldKey,
+    ...rest
+}) => {
+    const { ids: initialIds, data: initialData } = record
+        ? getDataAndIds(record, source, fieldKey)
```

Comment:
getDataAndIds is computationally intensive. To avoid running it at each render, I'd do the calculation inside a callback passed to useState(), as follows:

```diff
- const { ids: initialIds, data: initialData } = record ? getDataAndIds(record, source, fieldKey)
- const [ids, setIds] = useState(initialIds);
+ const [ids, setIds] = useState(() => {
+   const list = get(record, source);
+   if (!list) return [];
+   return fieldKey ? list.map(item => item[fieldKey]) : list.map(JSON.stringify);
+ });
```

Reply:
Done!